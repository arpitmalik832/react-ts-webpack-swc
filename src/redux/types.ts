import {
  BaseQueryFn,
  TypedUseQueryStateResult,
} from '@reduxjs/toolkit/query/react';
import { AxiosError, AxiosInstance } from 'axios';

interface SampleApiResponse {
  data: string[];
}

export interface BaseQueryParams {
  axiosInstance: AxiosInstance;
  url: string;
}

export type SampleQueryResponse = TypedUseQueryStateResult<
  SampleApiResponse,
  AxiosError,
  BaseQueryFn<BaseQueryParams>
>;
