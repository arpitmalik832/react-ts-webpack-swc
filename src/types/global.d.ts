interface Window {
  backPress: () => void;
}

declare module '*.svg' {
  export const ReactComponent: React.FunctionComponent<
    React.SVGProps<SVGSVGElement>
  >;
  const src: string;
  export default src;
}

declare module '*.png' {
  const value: string;
  export default value;
}

declare module '*.jpg' {
  const value: string;
  export default value;
}

declare module '*.jpeg' {
  const value: string;
  export default value;
}

declare module '*.gif' {
  const value: string;
  export default value;
}

declare module '*.bmp' {
  const value: string;
  export default value;
}

declare module '*.tiff' {
  const value: string;
  export default value;
}

declare module '*.module.scss' {
  const content: { [className: string]: string };
  export default content;
}
