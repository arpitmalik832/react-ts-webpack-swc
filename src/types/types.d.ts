import type {
  InternalAxiosRequestConfig,
  AxiosResponse,
  AxiosInstance,
} from 'axios';
import type { configureStore } from '@reduxjs/toolkit';

interface VoidFunction extends VoidFunction {
  (...args: any[]): void;
}

interface ComponentWithSuspenseProps {
  component: React.ReactNode;
  fallback?: React.ReactNode;
}

interface HtmlContentProps {
  title: string;
  description: string;
}

interface ReduxProviderProps {
  children?: React.ReactNode;
  store: ReturnType<typeof configureStore>;
}

interface IconProps {
  name: string;
}

type AbortControllers = Record<string, AbortController>;

interface RequestMetadata {
  startTime: Date;
  endTime: Date;
  responseTime: number | bigint;
}

interface AppRedux {
  theme: string;
}

interface ApisRedux {
  api1Host: string;
  api1Headers: Record<string, string>;
  api1AxiosInstance?: AxiosInstance;
}

interface NavigationRedux {
  stack: Array<VoidFunction>;
}

interface ReduxState {
  app: AppRedux;
  apis: ApisRedux;
  navigation: NavigationRedux;
}

type ColorTheme = 'light' | 'dark';
type ColorType = 'background';
type ColorInnerType = 'primary';
type PrimitiveColor = 'black' | 'white';
type ColorSemanticLabel = '50' | '100' | '900';
type HexColor = `#${string & { length: 6 | 8 }}`;
type TypographyScale = 'h1';
type TypographyScaleValue = '*px';
type TypographyWeight = 'bold';
type TypographyWeightValue = number;

interface DesignTokens {
  'color-primitives': Record<
    PrimitiveColor,
    Record<
      ColorSemanticLabel,
      {
        type: 'value';
        color: HexColor;
      }
    >
  >;
  'color-semantics': Record<
    Theme,
    Record<
      Type,
      Record<
        InnerType,
        Record<
          ColorSemanticLabel,
          {
            type: 'value';
            color: HexColor;
          }
        >
      >
    >
  >;
  typography: {
    scale: Record<
      TypographyScale,
      Record<
        'font-size' | 'line-height' | 'letter-spacing',
        Record<'value', TypographyScaleValue>
      >
    >;
    weight: Record<TypographyWeight, Record<'value', TypographyWeightValue>>;
  };
}

type MQEventListener = (this: MediaQueryList, ev: MediaQueryListEvent) => void;

type EventListener = (this: Window, ev: Event) => void;

type BeforeUnloadEventListener = (this: Window, ev: BeforeUnloadEvent) => void;
