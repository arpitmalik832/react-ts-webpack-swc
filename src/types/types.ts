import type {
  AxiosInstance,
} from 'axios';

export type AllParams =
  | string
  | string[]
  | object
  | object[]
  | number
  | number[]
  | bigint
  | bigint[]
  | boolean
  | boolean[];

export interface VoidFunctionWithParams extends VoidFunction {
  (...args: AllParams[]): void;
}

export interface KeyValuePair<T> {
  key: string;
  value: T | Record<string, T>;
}

export interface AppRedux {
  theme: string;
}

export interface NavigationRedux {
  stack: VoidFunctionWithParams[];
}

export interface APIData {
  host: string;
  headers: Record<string, string>;
  axiosInstance: AxiosInstance;
}

export type ApisRedux = APIData[];

export interface ReduxState {
  app: AppRedux;
  apis: ApisRedux;
  navigation: NavigationRedux;
}