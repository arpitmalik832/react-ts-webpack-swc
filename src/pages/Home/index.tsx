import {
  useBackPress,
  Button,
  log,
} from '@arpitmalik832/react-ts-rollup-library';
import { useSelector } from 'react-redux';
import { useEffect } from 'react';

import ButtonV2 from '../../components/atoms/Button';
import { ReactComponent as ReactIcon } from '../../assets/icons/react.svg';
import { ApisRedux, ReduxState } from '../../types/types';
import { useFetchDataQuery } from '../../redux/queries/sampleQuery';
import { SampleQueryResponse } from '../../redux/types';

function Home() {
  const apis = useSelector<ReduxState, ApisRedux>(state => state.apis);

  useBackPress();
  const { data, isLoading, isError } = useFetchDataQuery<SampleQueryResponse>(
    apis[0].axiosInstance,
  );

  useEffect(() => {
    log({ isLoading, data, isError });
  }, [isLoading, data, isError]);

  return (
    <div>
      Home
      <Button />
      <ButtonV2 />
      <ReactIcon />
    </div>
  );
}

export default Home;
