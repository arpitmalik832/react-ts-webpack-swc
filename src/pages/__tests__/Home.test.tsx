/**
 * Home Page unit tests.
 * @file This file is saved as `Home.test.jsx`.
 */
import { cleanup, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import {
  ReduxProvider,
  log,
  useBackPress,
} from '@arpitmalik832/react-ts-rollup-library';
import {
  configureStore,
  createSlice,
  EnhancedStore,
  PayloadAction,
  SliceCaseReducers,
  SliceSelectors,
} from '@reduxjs/toolkit';
import axios from 'axios';
import {
  sampleQuery,
  useFetchDataQuery,
} from '../../redux/queries/sampleQuery';

import Home from '../Home';
import { APIData, ApisRedux } from '../../types/types';

// Mock the library components and hooks
jest.mock('@arpitmalik832/react-ts-rollup-library', () => ({
  __esModule: true,
  ...jest.requireActual<object>('@arpitmalik832/react-ts-rollup-library'),
  log: jest.fn(),
  useBackPress: jest.fn(),
  Button: jest.fn(() => <div data-testid="library-button">Library Button</div>),
}));

// Mock the custom Button component
jest.mock('../../components/atoms/Button', () => ({
  __esModule: true,
  default: jest.fn(() => <div data-testid="custom-button">Custom Button</div>),
}));

// Mock the React icon
jest.mock('../../assets/icons/react.svg', () => ({
  ReactComponent: () => <div data-testid="react-icon">React Icon</div>,
}));

jest.mock('../../redux/queries/sampleQuery', () => ({
  ...jest.requireActual<object>('../../redux/queries/sampleQuery'),
  useFetchDataQuery: jest.fn(),
}));

describe('Home Page Tests', () => {
  let store: EnhancedStore<{
    apis: ApisRedux;
  }>;
  // Mock the query hook before rendering
  const mockQueryData = {
    data: { testData: 'test' },
    isLoading: false,
    isError: false,
  };

  beforeEach(() => {
    // Import and mock the query directly
    (useFetchDataQuery as jest.Mock).mockReturnValue(mockQueryData);

    // Create a slice for apis
    const apisSlice = createSlice<
      ApisRedux,
      SliceCaseReducers<ApisRedux>,
      string,
      SliceSelectors<ApisRedux>,
      string
    >({
      name: 'apis',
      initialState: [
        {
          host: 'no-url',
          headers: { x: 'a' },
          axiosInstance: axios.create({ baseURL: 'https://abc.com' }),
        },
      ],
      reducers: {
        addNewApiData: (state, action: PayloadAction<APIData>) => [
          ...state,
          action.payload,
        ],
      },
    });

    // Configure store before each test
    store = configureStore({
      reducer: {
        apis: apisSlice.reducer,
        [sampleQuery.reducerPath]: sampleQuery.reducer,
      },
      middleware: getDefault =>
        getDefault({
          serializableCheck: {
            ignoredActions: ['apis/addNewApiData'],
            ignoredPaths: ['apis', sampleQuery.reducerPath],
          },
        }).concat(sampleQuery.middleware),
    });
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  const renderComponent = () =>
    render(
      <ReduxProvider store={store}>
        <Home />
      </ReduxProvider>,
    );

  it('renders without crashing', () => {
    const { container } = renderComponent();
    expect(container).toBeTruthy();
  });

  it('renders all required components', () => {
    renderComponent();

    // Check for text content
    expect(screen.getByText('Home')).toBeInTheDocument();

    // Check for buttons
    expect(screen.getByTestId('library-button')).toBeInTheDocument();
    expect(screen.getByTestId('custom-button')).toBeInTheDocument();

    // Check for React icon
    expect(screen.getByTestId('react-icon')).toBeInTheDocument();
  });

  it('initializes with useBackPress hook', () => {
    renderComponent();

    expect(useBackPress).toHaveBeenCalled();
  });

  it('matches snapshot', () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });

  it('handles API state correctly', () => {
    renderComponent();

    expect(log).toHaveBeenCalledWith(mockQueryData);
  });

  it('renders correctly with loading state', () => {
    // Mock the useFetchDataQuery hook result for loading state
    (useFetchDataQuery as jest.Mock).mockReturnValue({
      data: undefined,
      isLoading: true,
      isError: false,
    });

    renderComponent();

    expect(log).toHaveBeenCalledWith({
      isLoading: true,
      data: undefined,
      isError: false,
    });
  });
});
