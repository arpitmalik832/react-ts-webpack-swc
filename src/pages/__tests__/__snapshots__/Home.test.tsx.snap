// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Tests matches snapshot 1`] = `
<div>
  <div>
    Home
    <div
      data-testid="library-button"
    >
      Library Button
    </div>
    <div
      data-testid="custom-button"
    >
      Custom Button
    </div>
    <div
      data-testid="react-icon"
    >
      React Icon
    </div>
  </div>
</div>
`;
